core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java

```diff
@@ -177,15 +181,17 @@ public ParsedQuery parse(QueryParseContext context, BytesReference source) {
         }
     }
 
-    public ParsedQuery parse(String source) throws QueryParsingException {
+    public ParsedQuery parse(String source) throws QueryShardException, QueryShardException {
```

Comment:
The whole class maybe? There are so many parse methods here that do the whole roundtrip.